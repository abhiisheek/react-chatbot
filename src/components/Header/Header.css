@import '../_variables.css';
@import '../_colors.css';
@import '../_defaults.css';

.header {
  @apply --chat-win-header;
  font-family: var(--preferred-font);
  cursor: pointer;
}

.minimiseIcon {
  @apply --chat-win-header-minimise-icon;
  cursor: pointer;
}

.strechIcon {
  @apply --chat-win-header-strech-icon;
  cursor: pointer;
}

.restoreIcon {
  @apply --chat-win-header-restore-icon;
  cursor: pointer;
}

.hideStretchIcon {
  display: none;
}

.backIcon {
  @apply --chat-win-header-back-icon;
  cursor: pointer;
}

.hideIcon {
  display: none;
}

.defaultCursor {
  cursor: default;
}
