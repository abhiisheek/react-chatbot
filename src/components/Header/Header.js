import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import classNames from 'classnames';

import styles from './Header.css';

/** Header component for the Chat window */
class Header extends PureComponent {
  handleOnMinimise = event => {
    event.stopPropagation();
    this.props.onMinimise(0);
  };

  handleOnMaximise = event => {
    event.stopPropagation();
    if (this.props.isDocked) {
      this.props.onMaximise(1);
    }
  };

  handleOnStretch = event => {
    event.stopPropagation();
    this.props.onStretch(2);
  };

  handleOnRestore = event => {
    event.stopPropagation();
    this.props.onRestore(1);
  };

  handleOnBack = event => {
    event.stopPropagation();
    this.props.onBack();
  };

  render() {
    const {
      title,
      isDocked,
      showBackIcon,
      windowState,
      isZoomSupported
    } = this.props;
    return (
      <div
        className={classNames(styles.header, {
          [`${styles.defaultCursor}`]: !isDocked
        })}
        onClick={this.handleOnMaximise}
      >
        <div
          className={classNames(styles.backIcon, {
            [`${styles.hideIcon}`]: !showBackIcon || isDocked
          })}
          onClick={this.handleOnBack}
        >
          {'←'}
        </div>
        {title}
        <div
          className={classNames(styles.minimiseIcon, {
            [`${styles.hideIcon}`]: isDocked
          })}
          onClick={this.handleOnMinimise}
        >
          {'–'}
        </div>
        <div
          className={classNames(styles.strechIcon, {
            [`${styles.hideIcon}`]:
              !isZoomSupported || (isDocked || windowState === 2)
          })}
          onClick={this.handleOnStretch}
        >
          {'✠'}
        </div>
        <div
          className={classNames(styles.restoreIcon, {
            [`${styles.hideIcon}`]:
              !isZoomSupported || (isDocked || windowState === 1)
          })}
          onClick={this.handleOnRestore}
        >
          {'❐'}
        </div>
      </div>
    );
  }
}

Header.displayName = 'Header';

Header.propTypes = {
  /** Title for the chat window */
  title: PropTypes.string.isRequired,
  /** Handler to handle minimise of chat window */
  onMinimise: PropTypes.func.isRequired,
  /** Handler to handle maximise of chat window */
  onMaximise: PropTypes.func.isRequired,
  /** Handler to going back to categories selection view from chatting section minimise of chat window */
  onBack: PropTypes.func.isRequired,
  /** Indicates the present state of chat window like minimise/maximise/zoom... */
  windowState: PropTypes.number.isRequired,
  /** Indicates whether the zoom capability needs to be supported by the chat window */
  isZoomSupported: PropTypes.bool,
  /** Handler to handle stretch/zoom of chat window. Required when isZoomSupported is set. */
  onStretch: PropTypes.func,
  /** Handler to handle restoring to normal size(zoom-out) of chat window. Required when isZoomSupported is set. */
  onRestore: PropTypes.func,
  /** Indicates whether chat window is minimised or not */
  isDocked: PropTypes.bool,
  /** Indicator to show back icon or not */
  showBackIcon: PropTypes.bool
};

Header.defaultProps = {
  isDocked: true,
  showBackIcon: false,
  isZoomSupported: false
};

export default Header;
