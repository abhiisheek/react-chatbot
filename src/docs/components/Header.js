import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import styles from './header.css';

class Header extends PureComponent {
  render() {
    const { className } = this.props;
    return (
      <header className={`${styles.wrapper} ${className}`}>
        <a className={styles.banner} href="/#">
          React Chatbot Component Library
        </a>
        <nav>
          <ul>
            <li className={styles.nav_item}>
              <a
                className={styles.nav_achor}
                href="https://github.com/abhiisheek/react-chatbot"
              >
                Github
              </a>
            </li>
          </ul>
        </nav>
      </header>
    );
  }
}

Header.propTypes = {
  className: PropTypes.string
};

export default Header;
