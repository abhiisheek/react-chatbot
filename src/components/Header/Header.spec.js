import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Header from './Header';

let handleOnMinimise;
let handleOnMaxmise;
let handleOnBack;
let handleOnStretch;
let handleOnRestore;
const title = 'Chat Bot';
const isDocked = true;
const showBackIcon = false;

const create = (
  docked,
  showBackIcon,
  isZoomSupported = false,
  windowState = 0
) =>
  shallow(
    <Header
      title={title}
      onMinimise={handleOnMinimise}
      onMaximise={handleOnMaxmise}
      onStretch={handleOnStretch}
      onRestore={handleOnRestore}
      isDocked={docked}
      isZoomSupported={isZoomSupported}
      onBack={handleOnBack}
      showBackIcon={showBackIcon}
      windowState={windowState}
    />
  );

describe('Header', () => {
  beforeEach(() => {
    handleOnMinimise = jest.fn();
    handleOnMaxmise = jest.fn();
    handleOnBack = jest.fn();
    handleOnStretch = jest.fn();
    handleOnRestore = jest.fn();
  });

  it('Should have displayName set on the component', () => {
    expect(Header.displayName).toBeDefined();
  });

  it('Snapshot test', () => {
    const snapshot = renderer
      .create(
        <Header
          title={title}
          onMinimise={handleOnMinimise}
          onMaximise={handleOnMaxmise}
          isDocked={isDocked}
          onBack={handleOnBack}
          showBackIcon={showBackIcon}
          windowState={0}
        />
      )
      .toJSON();
    expect(snapshot).toMatchSnapshot();
  });

  it('Should not display minimise icon when state is docked', () => {
    const wrapper = create(isDocked);
    expect(wrapper.find('.minimiseIcon').hasClass('hideIcon')).toBeTruthy();
    expect(wrapper.hasClass('defaultCursor')).toBeFalsy();
  });

  it('Should display minimise icon when state is not docked', () => {
    const wrapper = create(!isDocked, undefined, undefined, 1);
    expect(wrapper.find('.minimiseIcon').hasClass('hideIcon')).toBeFalsy();
    expect(wrapper.hasClass('defaultCursor')).toBeTruthy();
  });

  it('Should not display back icon when showBackIcon is unset', () => {
    const wrapper = create(isDocked, false);
    expect(wrapper.find('.backIcon').hasClass('hideIcon')).toBeTruthy();
  });

  it('Should not display back icon when state is not docked', () => {
    const wrapper = create(isDocked, true);
    expect(wrapper.find('.backIcon').hasClass('hideIcon')).toBeTruthy();
  });

  it('Should display back icon when showBackIcon is set', () => {
    const wrapper = create(!isDocked, true);
    expect(wrapper.find('.backIcon').hasClass('hideIcon')).toBeFalsy();
  });

  it('Should display zoom/strech icon when chat window is not minimised or zoomed', () => {
    const wrapper = create(false, true, true, 1);
    expect(wrapper.find('.strechIcon').hasClass('hideIcon')).toBeFalsy();
    expect(wrapper.find('.restoreIcon').hasClass('hideIcon')).toBeTruthy();    
  });

  it('Should not display zoom/strech icon when zoom is not supported', () => {
    const wrapper = create(false, true, false, 1);
    expect(wrapper.find('.strechIcon').hasClass('hideIcon')).toBeTruthy();
  });

  it('Should display restore icon when chat window is zoomed', () => {
    const wrapper = create(false, true, true, 2);
    expect(wrapper.find('.restoreIcon').hasClass('hideIcon')).toBeFalsy();
    expect(wrapper.find('.strechIcon').hasClass('hideIcon')).toBeTruthy();    
  });

  it('Should not display restore icon when zoom is not supported', () => {
    const wrapper = create(false, true, false, 1);
    expect(wrapper.find('.restoreIcon').hasClass('hideIcon')).toBeTruthy();
  });

  it('Should display given titel', () => {
    const wrapper = create(isDocked);
    expect(
      wrapper
        .find('.header')
        .text()
        .includes(title)
    ).toBeTruthy();
  });

  it('Should call hadleOnMinimise', () => {
    const event = {
      stopPropagation: jest.fn()
    };
    const wrapper = create(false);
    wrapper.find('.minimiseIcon').simulate('click', event);
    expect(handleOnMinimise).toHaveBeenCalled();
    expect(event.stopPropagation).toHaveBeenCalled();
    expect(event.stopPropagation).toHaveBeenCalledTimes(1);
  });

  it('Should call handleOnMaxmise', () => {
    const event = { stopPropagation: jest.fn() };
    const wrapper = create(isDocked);
    wrapper.find('.header').simulate('click', event);
    expect(handleOnMaxmise).toHaveBeenCalled();
    expect(event.stopPropagation).toHaveBeenCalledTimes(1);
    expect(event.stopPropagation).toHaveBeenCalled();
  });

  it('Should call handleOnBack', () => {
    const event = { stopPropagation: jest.fn() };
    const wrapper = create(isDocked, true);
    wrapper.find('.backIcon').simulate('click', event);
    expect(handleOnBack).toHaveBeenCalled();
    expect(event.stopPropagation).toHaveBeenCalledTimes(1);
    expect(event.stopPropagation).toHaveBeenCalled();
  });

  it('Should call handleOnRestore', () => {
    const event = { stopPropagation: jest.fn() };
    const wrapper = create(isDocked, true, true, 2);
    wrapper.find('.restoreIcon').simulate('click', event);
    expect(handleOnRestore).toHaveBeenCalled();
    expect(event.stopPropagation).toHaveBeenCalledTimes(1);
    expect(event.stopPropagation).toHaveBeenCalled();
  });

  it('Should call handleOnStretch', () => {
    const event = { stopPropagation: jest.fn() };
    const wrapper = create(isDocked, true, true, 1);
    wrapper.find('.strechIcon').simulate('click', event);
    expect(handleOnStretch).toHaveBeenCalled();
    expect(event.stopPropagation).toHaveBeenCalledTimes(1);
    expect(event.stopPropagation).toHaveBeenCalled();
  });

  it('Should not call handleOnMaxmise when window is docked', () => {
    const event = { stopPropagation: jest.fn() };
    const wrapper = create(!isDocked);
    wrapper.find('.header').simulate('click', event);
    expect(handleOnMaxmise).not.toHaveBeenCalled();
    expect(event.stopPropagation).toHaveBeenCalledTimes(1);
    expect(event.stopPropagation).toHaveBeenCalled();
  });
});
