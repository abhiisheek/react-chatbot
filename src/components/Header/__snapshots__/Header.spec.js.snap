// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Snapshot test 1`] = `
<div
  className="header"
  onClick={[Function]}
>
  <div
    className="backIcon hideIcon"
    onClick={[Function]}
  >
    ←
  </div>
  Chat Bot
  <div
    className="minimiseIcon hideIcon"
    onClick={[Function]}
  >
    –
  </div>
  <div
    className="strechIcon hideIcon"
    onClick={[Function]}
  >
    ✠
  </div>
  <div
    className="restoreIcon hideIcon"
    onClick={[Function]}
  >
    ❐
  </div>
</div>
`;
